torchmultimodal/architectures/clip.py

```diff
@@ -20,32 +22,38 @@ class CLIPArchitecture(nn.Module):
     encoders, while the loss is implemented in ContrastiveLossWithTemperature.
 
 
-    Args:   vision_encoder (nn.Module): Instantiated vision encoder.
-                See e.g. ResNetForCLIP class.
-            text_encoder (nn.Module): Instantiated text encoder.
-                See CLIPTextEncoder class.
+    Args:   encoders (nn.ModuleDict): Dict of instantiated encoders, keyed by modality.
+                E.g. {"vision": ResNetForCLIP(), "text": CLIPTextEncoder()}
 
-    Inputs: image (Tensor): Tensor containing image features.
-            text (Tensor): Tensor containing text features.
+    Inputs: modalities (Dict[str, Tensor]): Dict of Tensor features, keyed by modality.
+                Must contain one entry for every modality in ``encoders``.
+
+    Output: CLIPOutput object with fields ``{modality}_embeddings`` for every modality
+                in ``encoders``.
     """
 
     def __init__(
         self,
-        vision_encoder: nn.Module,
-        text_encoder: nn.Module,
+        encoders: nn.ModuleDict,
     ):
         super().__init__()
-        self.vision_encoder = vision_encoder
-        self.text_encoder = text_encoder
+        self.encoders = nn.ModuleDict({k: encoders[k] for k in sorted(encoders.keys())})
 
     def forward(
         self,
-        image: torch.Tensor,
-        text: torch.Tensor,
+        modalities: Dict[str, torch.Tensor],
     ) -> Dict[str, torch.Tensor]:
-
-        img_embeddings = self.vision_encoder(image)
-        text_embeddings = self.text_encoder(text)
-        img_embeddings = F.normalize(img_embeddings)
-        text_embeddings = F.normalize(text_embeddings)
-        return {"image": img_embeddings, "text": text_embeddings}
+        embeddings = {}
+        for key, encoder in self.encoders.items():
+            assert key in modalities, f"{key} missing in input"
+            embeddings[f"{key}_embeddings"] = F.normalize(encoder(modalities[key]))
```

Comment:
Random question: do we definitely want normalize in the architecture? (I know it's already here now so not necessarily relevant to your PR, but this is another piece we could consider making the encoder's responsibility, similar to projections as we discussed)

Reply:
The L_p normalization puts the encodings for different modalities within the same scale `(0, 1)`? I suppose that's an essential step before computing the cosine similarity. The CLIP paper also has this step: https://arxiv.org/pdf/2103.00020.pdf.